File: c/data_structures/kth_smallest_in_BST.c
/*
 * This program prints the kth smallest number in a BST
 *
 * Author: @pyav
 */

#include <stdio.h>
#include <stdlib.h>

#define SUCCESS 0
#define FAILURE 1

/* The node structure */
struct node {
  int data;
  struct node *left;
  struct node *right;
};

/* Create a new node */
struct node *new_node(int data)
{
  struct node *one_node = NULL;

  one_node = (struct node *) malloc(sizeof (struct node));
  if (one_node != NULL) {
    one_node->data = data;
    one_node->left = NULL;
    one_node->right = NULL;
  }

  return one_node;
}

/* Insert a node */
struct node *insert_node(struct node *node, int data)
{
  if (node == NULL)
    return new_node(data);

  if (data < node->data)
    node->left = insert_node(node->left, data);
  else if (data > node->data)
    node->right = insert_node(node->right, data);

  return node;
}

/* Print tree inorder */
void print_tree_inorder(struct node *node)
{
  if (node != NULL) {
    print_tree_inorder(node->left);
    printf("%d ", node->data);
    print_tree_inorder(node->right);
  }
}

static int count = 0;

/* Find the kth smallest element */
void find_kth_smallest_elem(struct node *node, int k)
{
  if (node == NULL)
    return;

  find_kth_smallest_elem(node->left, k);
  count++;

  if (k == count) {
    printf("\n%dth smallest value in the BST is: %d\n", k, node->data);
    exit(EXIT_SUCCESS);
  }

  find_kth_smallest_elem(node->right, k);
}

/* Cleanup */
void free_all(struct node **node)
{
  if (*node == NULL)
    return;

  free_all(&(*node)->left);
  free_all(&(*node)->right);
  free(*node);
  *node = NULL;
}

/* Main driver program */
int main(void)
{
  int k = -1;
  struct node *root = NULL;

  /* Create the tree */
  root = insert_node (root, 50);
  insert_node (root, 10);
  insert_node (root, 55);
  insert_node (root, 70);
  insert_node (root, 40);
  insert_node (root, 30);

  print_tree_inorder(root);

  printf("\nEnter the value of k: ");
  scanf("%d", &k);

  /* Find the kth smallest element */
  find_kth_smallest_elem(root, k);

  /* Cleanup */
  free_all(&root);

  return SUCCESS;
}

File: src/leetcode/KthSmallestInBST.java
/**
 * https://leetcode.com/problems/kth-smallest-element-in-a-bst/
 *
 * Output:
 * ------
 * 1
 * 2
 * 3
 * 4
 * 5
 * 6
 */

public class KthSmallestInBST {

    static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    private int ht = 0;
    private int reqVal = 0;

    public int kthSmallest(TreeNode root, int k) {
        if (null == root) {
            return reqVal;
        }
        kthSmallest(root.left, k);
        ht++;
        if (ht == k) {
            reqVal = root.val;
        }
        kthSmallest(root.right, k);
        return reqVal;
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(5);
        root.left = new TreeNode(3);
        root.right = new TreeNode(6);
        root.left.right = new TreeNode(4);
        root.left.left = new TreeNode(2);
        root.left.left.left = new TreeNode(1);
        System.out.println(new KthSmallestInBST().kthSmallest(root, 1));
        System.out.println(new KthSmallestInBST().kthSmallest(root, 2));
        System.out.println(new KthSmallestInBST().kthSmallest(root, 3));
        System.out.println(new KthSmallestInBST().kthSmallest(root, 4));
        System.out.println(new KthSmallestInBST().kthSmallest(root, 5));
        System.out.println(new KthSmallestInBST().kthSmallest(root, 6));
    }

}

